### PR TITLE
Cm 102 fix objectives delete error

### PR DESCRIPTION
**Pull request by dcmcand**
**Description of change**
Fixes bug where objectives from saved goals were being provisioned in a new report when the saved goal was added. If you tried to delete these goals and then save the report, it caused a 500 response because it would try to delete an objective that was being used by another report.

Thanks @rahearn for figuring out how to reproduce the bug and @jasalisbury for pinning down where exactly the fix should go.

The solution had 2 parts.
1. Change the `removeUnusedObjectivesGoalsFromReport` function in `src/services/goals.js` to not delete objectives that were used by other reports.
2. Update the `ActivityReports` model so the goals virtual field only returns objectives that belong to that report instead of all objectives associated with a report.

**How to test**
Checkout the main branch and complete a activity report, submit it, and approve it.
Create a second activity report using the same grantee from the first report.
Add the goal that was saved in the first report. Save the report and refresh the page.
You will see objectives from the first and second report.
Save the draft again.
At this point, stop  the app, checkout this branch and start the app again.
Refresh the page in the browser and you will still see both objectives.
Delete the objective from the completed report and save the report.
It should save successfully.
Create a third report still running on this branch. Use the same grantee and add the goal from the first report.
Refresh the page. You should only see the objective you just added, not both of them.

**Issue(s)**
* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-102

**Checklist**

- [X] Meets issue criteria
- [X] Code tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

File: frontend/yarn.lock [hunk 1]
```diff
@@ -10907,9 +10902,9 @@ sshpk@^1.7.0:
     tweetnacl "~0.14.0"
 
 ssri@^6.0.1:
-  version "6.0.1"
-  resolved "https://registry.yarnpkg.com/ssri/-/ssri-6.0.1.tgz#2a3c41b28dd45b62b63676ecb74001265ae9edd8"
-  integrity sha512-3Wge10hNcT1Kur4PDFwEieXSCMCJs/7WvSACcrMYrNp+b8kDL1/0wJch5Ni2WrtwEa2IO8OsVfeKIciKCDx/QA==
+  version "6.0.2"
+  resolved "https://registry.yarnpkg.com/ssri/-/ssri-6.0.2.tgz#157939134f20464e7301ddba3e90ffa8f7728ac5"
+  integrity sha512-cepbSq/neFK7xB6A50KHN0xHDotYzq58wWCa5LeWqnPrHG8GzfEjO/4O8kpmcGW+oaxkvhEJCWgbgNk4/ZV93Q==
   dependencies:
```
Comment (dcmcand) [inline frontend/yarn.lock:10908]: Upgrade ssri to 6.0.2 to fix a security vulnerability

File: src/models/activityReport.js [hunk 2]
```diff
@@ -171,7 +171,22 @@ export default (sequelize, DataTypes) => {
       type: DataTypes.VIRTUAL,
       get() {
         const objectives = this.objectives || [];
-        return uniqBy(objectives.map((o) => o.goal), 'id');
+        const goalsArray = objectives.map((o) => o.goal);
+        const goals = uniqWith(goalsArray, isEqual);
+
```
Comment (dcmcand) [inline src/models/activityReport.js:176]: This prevents dupes from showing up in the goal list

File: src/services/activityReports.js [hunk 3]
```diff
@@ -229,10 +229,6 @@ export function activityReportById(activityReportId) {
         include: [{
           model: Goal,
           as: 'goal',
-          include: [{
-            model: Objective,
-            as: 'objectives',
-          }],
         }],
```
Comment (dcmcand) [inline src/services/activityReports.js:232]: This is what was responsible for all of the objectives showing up.

File: src/services/goals.js [hunk 4]
```diff
@@ -83,9 +83,18 @@ async function removeUnusedObjectivesGoalsFromReport(reportId, currentGoals, tra
...
+      const notCurrent = !currentObjectiveIds.includes(id);
+      const activityReportObjectives = await ActivityReportObjective
+        .findAll({ where: { objectiveId: id } });
+      const lastInstance = activityReportObjectives.length <= 1;
+      if (notCurrent && lastInstance) {
+        objectiveIdsToRemove.push(id);
+      }
+    }),
+  );
   await removeActivityReportObjectivesFromReport(reportId, transaction);
```
Comment (dcmcand) [inline src/services/goals.js:98]: This is what was actually causing the 500's when people tried to remove the objectives that were added from previous reports. This filters them and only deletes objectives that are not in use by other reports.

File: src/services/goals.js [hunk 5]
```diff
@@ -83,9 +83,18 @@ async function removeUnusedObjectivesGoalsFromReport(reportId, currentGoals, tra
   const previousObjectiveIds = previousObjectives.map((ro) => ro.objectiveId);
 
   const goalIdsToRemove = previousGoalIds.filter((id) => !currentGoalIds.includes(id));
-  const objectiveIdsToRemove = previousObjectiveIds
-    .filter((id) => !currentObjectiveIds.includes(id));
-
+  const objectiveIdsToRemove = [];
+  Promise.all(
```
Comment (jasalisbury) [inline src/services/goals.js:87]: Do we need to await this `promise.all`?

**Review by jasalisbury — APPROVED** [review]
Looks good! Just have one question

File: src/services/goals.js (hunk 5 above)
Comment (dcmcand) [reply src/services/goals.js:87]: It hasn't been a problem, but I will add it just to be safe.

**Comment by kryswisnaskas** [comment]
Works as advertised!

It appears though that the removed objectives that don't belong to any other reports, persist in the database. Could we remove the unused objectives? To see this:

- create a report and add a goal with some objectives
- save the report
- submit that report and approve it
- create a second report for the same grantee
- select the previously created goal, add an objective "test" and save
- remove the goal from the report
- re-add the same goal
- create a new objective and save
- observe that the previous objective "test" still persist in the database

**Comment by dcmcand** [comment]
> Works as advertised!
> 
> It appears though that the removed objectives that don't belong to any other reports, persist in the database. Could we remove the unused objectives? To see this:
> 
>     * create a report and add a goal with some objectives
> 
>     * save the report
> 
>     * submit that report and approve it
> 
>     * create a second report for the same grantee
> 
>     * select the previously created goal, add an objective "test" and save
> 
>     * remove the goal from the report
> 
>     * re-add the same goal
> 
>     * create a new objective and save
> 
>     * observe that the previous objective "test" still persists in the database

I think adding a nightly worker job that cleans up any objectives not associated with an Activity Report would be pretty easy to do. I think that could be a separate ticket unless you feel strongly that it needs added here.

**Comment by kryswisnaskas** [comment]
There is also an error, which might be fixed once we remove the unused objectives. If not, we could create a separate ticket if it's too much for this PR. It shows up when removing a goal and saving. Steps to reproduce:

- create a report add a goal with an objective and save
- remove the created goal and try to save
- notice "Unable to save activity report" with
`error:   SERVICE:ACTIVITY_REPORTS id: 13 Sequelize error SequelizeForeignKeyConstraintError: update or delete on table "Goals" violates foreign key constraint "Objectives_goalId_fkey" on table "Objectives"`

**Comment by dcloud** [comment]
I think the code here fixes the 500 issue, and I've seen other issues related to objectives while looking at this. 

If I `SELECT * FROM "Objectives" WHERE title = ''` locally, I see a number of "empty" objectives, and I think this occurs due to autosave running when empty objective forms are open. I had a window open with some empty objective forms, and later input information, and saw that a record that had been empty is no longer.

I also saw that if I had saved a report with two Goals, and removed one, then the other goal's objectives would be removed. I think that I will try to tackle that in https://ocio-jira.acf.hhs.gov/browse/TTAHUB-87 which maybe implies that issue?

**Comment by dcmcand** [comment]
> There is also an error, which might be fixed once we remove the unused objectives. If not, we could create a separate ticket if it's too much for this PR. It shows up when removing a goal and saving. Steps to reproduce:
> 
>     * create a report add a goal with an objective and save
> 
>     * remove the created goal and try to save
> 
>     * notice "Unable to save activity report" with
>       `error:   SERVICE:ACTIVITY_REPORTS id: 13 Sequelize error SequelizeForeignKeyConstraintError: update or delete on table "Goals" violates foreign key constraint "Objectives_goalId_fkey" on table "Objectives"`

I cannot reproduce this. I am going to deploy into sandbox. Can you try it then and see if it still happens?

**Comment by kryswisnaskas** [comment]
> I think adding a nightly worker job that cleans up any objectives not associated with an Activity Report would be pretty easy to do. I think that could be a separate ticket unless you feel strongly that it needs added here.

If possible, it would be cleaner to be able to remove the objectives as the goals get removed.

**Comment by kryswisnaskas** [comment]
> I cannot reproduce this. I am going to deploy into sandbox. Can you try it then and see if it still happens?
I don't see this issue on sandbox 👍

**Comment by dcmcand** [comment]
Can you check the other issue too? Cause I am actually seeing those objectives getting deleted.

**Review by rahearn — APPROVED** [review]

**Comment by kryswisnaskas** [comment]
> Can you check the other issue too? Cause I am actually seeing those objectives getting deleted.
I just tried your steps locally and the orphaned objective was deleted as it should be.

I am having problems connecting to the database on staging (ssh) at the moment, but I rebuilt my local database and tried again on the latest version of this branch. Good news is that I don't see this behavior anymore. 🎉 

Thanks @dcloud for pointing out the other potential issues. As long as that's handled separately in 87, we should be good to go.

**Review by kryswisnaskas — APPROVED** [review]

**Pull request merged**